proposals/kapp-controller/004-dependency-resolution/README.md

```diff
@@ -0,0 +1,619 @@
+          constraints: 0.2.0
+      packageInstallRef:
+        name: other-pkg.pkg5.test.carvel.dev
+        namespace: ns1
+      conditions:
+      - type: FoundPackageInstall
+
+    - name: microservice-v2
+      wantedPackageRef:
+        refName: pkg5.test.carvel.dev
+        versionSelection:
+          constraints: 0.2.0
+      packageInstallRef:
+        name: pkg-demo.microservice-v2
+        namespace: default
+      conditions:
+      - type: OwnedPackageInstall
+
+    - name: optional-dependency
+      wantedPackageRef:
+        refName: pkg6.test.carvel.dev
+        versionSelection:
+          constraints: 0.6.0
+      conditions:
+      - type: UnknownPackage
+
+    - name: dep-with-wrong-version
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: "1.0.1" # one on the cluster is 0.1.1
+      packageInstallRef:
+        name: pkg-demo.pkg1.test.carvel.dev
+      conditions:
+      - type: FoundPackageInstall
+      - type: FoundPackageInstallRequiresDifferentVersion
+
+    - name: dep-with-unknown-package
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+      conditions:
+      - type: UnknownPackage
+
+    - name: dep-with-package-with-high-version
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: ">4.0.0" # cant calculate exact version in this system
+      conditions:
+      - type: UnknownPackageVersion
+
+```
+
+Example of a PackageInstall that is automatically created
+```yaml
+apiVersion: packaging.carvel.dev/v1alpha1
+kind: PackageInstall
+metadata:
+  name: pkg1.test.carvel.dev.1.0.1
+  annotations:
+    kapp-controller.carvel.dev/owner: "PackageInstall/pkg-demo"
+spec:
+    serviceAccountName: default-ns-sa
+  packageRef:
+    refName: pkg3.test.carvel.dev
+    versionSelection:
+      constraints: 1.0.0
+  values:
+    secretRef:
+      name: pkg3.dev-values
+...
+```
+
+#### Dependency resolution
+The dependency resolution will be done one PackageInstall at a time, basically each PackageInstall will only create the PackageInstalls that it directly depends on.
+
+ex:
+```
+When pkg depends on pkg1 and pkg3, while pkg1 depends on pkg2
+kapp-controller when it is processing pkg it will only
+create PackageInstalls for pkg1 and pkg3.
+
+pkg
+  -> pkg1
+      -> pkg2
+  -> pkg3
+```
+
+For future versions we can contemplate the possibility of each PackageInstall checking if the dependencies it has can be fulfilled with the Packages present in the Cluster.
+
+#### Use Cases
+
+##### Package depends on another Package
+- When the Package that match the dependency is already installed, no new installation is needed
+- When the dependent Pacakage is not installed, it will try to install it
+- When a dependent Package is already installed but the version does not fullfil the requirements of the current PackageInstall, kapp-controller will try to install the version that matches
```

Comment:
What happens to the older version, its upgraded or both versions are installed? It would be helpful to add some more thoughts around that (Given I assume we don't different versions of the same GK owned by multiple packages, or is this not the case?)

Reply:
kapp-controller will try to install the version that is required by this package, the existing installation will remain as is. I will add more details around this, thanks!